skcosmo/pcovr/pcovr.py

```diff
@@ -0,0 +1,350 @@
+
+            \\mathbf{P}_{TX} = \\mathbf{\\Lambda}_\\mathbf{\\tilde{C}}^{-\\frac{1}{2}}
+                                \\mathbf{U}_\\mathbf{\\tilde{C}}^T
+                                (\\mathbf{X}^T \\mathbf{X})^{\\frac{1}{2}}
+
+        .. math::
+
+            \\mathbf{P}_{TY} = \\mathbf{\\Lambda}_\\mathbf{\\tilde{C}}^{-\\frac{1}{2}}
+                               \\mathbf{U}_\\mathbf{\\tilde{C}}^T (\\mathbf{X}^T
+                               \\mathbf{X})^{-\\frac{1}{2}} \\mathbf{X}^T
+                               \\mathbf{Y}
+
+        """
+
+        Ct, iCsqrt = pcovr_covariance(
+            mixing=self.mixing,
+            X_proxy=X,
+            Y_proxy=Yhat,
+            rcond=self.tol,
+            return_isqrt=True,
+        )
+        Csqrt = np.linalg.inv(iCsqrt)
+
+        v, U = self._eig_solver(Ct)
+        S = v ** 0.5
+        S_inv = np.linalg.pinv(np.diagflat(S))
+
+        self.singular_values_ = S.copy()
+        self.explained_variance_ = (S ** 2) / (X.shape[0] - 1)
+        self.explained_variance_ratio_ = (
+            self.explained_variance_ / self.explained_variance_.sum()
+        )
+
+        self.pxt_ = np.linalg.multi_dot([iCsqrt, U, np.diagflat(S)])
+        self.ptx_ = np.linalg.multi_dot([S_inv, U.T, Csqrt])
+        self.pty_ = np.linalg.multi_dot([S_inv, U.T, iCsqrt, X.T, Yhat])
+
+    def _fit_structure_space(self, X, Yhat, W):
+        """
+        In sample-space PCovR, the projectors are determined by:
+
+        .. math::
+
+            \\mathbf{\\tilde{K}} = \\alpha \\mathbf{X} \\mathbf{X}^T +
+            (1 - \\alpha) \\mathbf{\\hat{Y}}\\mathbf{\\hat{Y}}^T
+
+        where
+
+        .. math::
+
+            \\mathbf{P}_{XT} = \\left(\\alpha \\mathbf{X}^T + (1 - \\alpha)
+                               \\mathbf{W} \\mathbf{\\hat{Y}}^T\\right)
+                               \\mathbf{U}_\\mathbf{\\tilde{K}}
+                               \\mathbf{\\Lambda}_\\mathbf{\\tilde{K}}^{-\\frac{1}{2}}
+
+        .. math::
+
+            \\mathbf{P}_{TX} = \\mathbf{\\Lambda}_\\mathbf{\\tilde{K}}^{-\\frac{1}{2}}
+                                \\mathbf{U}_\\mathbf{\\tilde{K}}^T \\mathbf{X}
+
+        .. math::
+
+            \\mathbf{P}_{TY} = \\mathbf{\\Lambda}_\\mathbf{\\tilde{K}}^{-\\frac{1}{2}}
+                               \\mathbf{U}_\\mathbf{\\tilde{K}}^T \\mathbf{Y}
+
+        """
+
+        Kt = pcovr_kernel_distance(mixing=self.mixing, X_proxy=X, Y_proxy=Yhat)
+
+        v, U = self._eig_solver(Kt)
+        S = v ** 0.5
+
+        P = (self.mixing * X.T) + (1.0 - self.mixing) * np.dot(W, Yhat.T)
+
+        self.singular_values_ = S.copy()
+        self.explained_variance_ = (S ** 2) / (X.shape[0] - 1)
+        self.explained_variance_ratio_ = (
+            self.explained_variance_ / self.explained_variance_.sum()
+        )
+
+        T = U @ np.diagflat(1 / S)
+
+        self.pxt_ = P @ T
+        self.pty_ = T.T @ Yhat
+        self.ptx_ = T.T @ X
+
+    def _eig_solver(self, matrix):
```

Comment:
A very similar function is needed in Sparse methods. It cannot be generalized and taken out somewhere in order not to duplicate the code?

Reply:
Sure, we could make a `utils` folder, what do you think?

Reply:
Yep, I think, it's a great idea

Reply:
Done!